components/sup/src/manager/user_config_watcher.rs

```diff
@@ -76,21 +79,26 @@ struct WorkerState {
 
 type ServiceName = String;
 pub struct UserConfigWatcher {
-    states: HashMap<ServiceName, WorkerState>,
+    // We use Arc/Mutex here, because this needs to be shareable
+    // across threads so we can remove watchers from futures; the
+    // Sender and Receiver members of WorkerState aren't themselves
+    // shareable.
+    states: Arc<Mutex<HashMap<ServiceName, WorkerState>>>,
 }
 
 impl UserConfigWatcher {
     pub fn new() -> Self {
         Self {
-            states: HashMap::new(),
+            states: Arc::new(Mutex::new(HashMap::new())),
         }
     }
 
     /// Adds a service to the User Config Watcher, thereby starting a watcher thread.
     pub fn add<T: Serviceable>(&mut self, service: &T) -> io::Result<()> {
         // It isn't possible to use the `or_insert_with` function here because it can't have a
         // return value, which we need to return the error from `Worker::run`.
-        if self.states.get(service.name()).is_none() {
+        let mut states = self.states.lock().unwrap();
```

Comment:
Any reason not to use `expect` in case of mutex poisoning here as we do elsewhere?

Reply:
I'll change it to be more consistent with our other usage.

It _does_ appear to be somewhat idiomatic usage in other Rust code I've been reading, though, where it's basically saying "A panic from thread X should propagate to the current thread Y".

`expect` gives better error messages, though.

Reply:
Force-pushed an amended commit using `expect` instead of `unwrap`.